File: CICD Agent/runtime/config/profiles.py
"""Profile loader.

Profiles live in `runtime/config/profiles.yaml` by default, but the user can
override with an env var `CICD_AGENT_PROFILES_PATH` to point at a per-machine
or per-repo file (e.g. `<repo>/.cicd-agent/profiles.yaml`).
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

DEFAULT_PROFILES_PATH = Path(__file__).with_name("profiles.yaml")


@dataclass(frozen=True)
class BuildSpec:
    command: str = ""


@dataclass(frozen=True)
class TestSpec:
    command: str = ""


@dataclass(frozen=True)
class AzureDevOpsSpec:
    organization: str = ""
    project: str = ""
    repository: str = ""
    default_target_branch: str = "main"
    pipeline_id: int | None = None


@dataclass(frozen=True)
class Profile:
    name: str
    description: str = ""
    languages: tuple[str, ...] = ()
    build: BuildSpec = field(default_factory=BuildSpec)
    test: TestSpec = field(default_factory=TestSpec)
    azure_devops: AzureDevOpsSpec = field(default_factory=AzureDevOpsSpec)
    ignored_globs: tuple[str, ...] = ()


def _resolve_path() -> Path:
    override = os.environ.get("CICD_AGENT_PROFILES_PATH")
    if override:
        p = Path(override)
        if p.exists():
            return p
    return DEFAULT_PROFILES_PATH


def _coerce_profile(name: str, raw: dict[str, Any]) -> Profile:
    build_raw = raw.get("build") or {}
    test_raw = raw.get("test") or {}
    ado_raw = raw.get("azure_devops") or {}
    return Profile(
        name=name,
        description=str(raw.get("description", "")),
        languages=tuple(raw.get("languages") or ()),
        build=BuildSpec(command=str(build_raw.get("command", ""))),
        test=TestSpec(command=str(test_raw.get("command", ""))),
        azure_devops=AzureDevOpsSpec(
            organization=str(ado_raw.get("organization", "")),
            project=str(ado_raw.get("project", "")),
            repository=str(ado_raw.get("repository", "")),
            default_target_branch=str(ado_raw.get("default_target_branch", "main")),
            pipeline_id=ado_raw.get("pipeline_id"),
        ),
        ignored_globs=tuple(raw.get("ignored_globs") or ()),
    )


def load_profiles(path: Path | None = None) -> dict[str, Profile]:
    """Load all profiles from YAML. Returns dict keyed by profile name."""
    target = path or _resolve_path()
    if not target.exists():
        return {"default": Profile(name="default")}
    with target.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f) or {}
    profiles_raw = data.get("profiles") or {}
    return {name: _coerce_profile(name, raw or {}) for name, raw in profiles_raw.items()}


def get_profile(name: str, path: Path | None = None) -> Profile:
    """Return a single profile by name, falling back to `default`."""
    profiles = load_profiles(path)
    if name in profiles:
        return profiles[name]
    if "default" in profiles:
        return profiles["default"]
    return Profile(name=name)

File: packages/core/src/profiles.ts
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import YAML from "yaml";
import { z } from "zod";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ─── YAML template profile schema (unchanged) ─────────────────────────────────

const BuildSpec = z.object({ command: z.string().default("") }).default({ command: "" });
const TestSpec = z.object({ command: z.string().default("") }).default({ command: "" });

const AzureDevOpsSpec = z
  .object({
    organization: z.string().default(""),
    project: z.string().default(""),
    repository: z.string().default(""),
    default_target_branch: z.string().default("main"),
    pipeline_id: z.number().int().nullable().default(null),
  })
  .default({});

const ProfileSchema = z.object({
  description: z.string().default(""),
  languages: z.array(z.string()).default([]),
  build: BuildSpec,
  test: TestSpec,
  azure_devops: AzureDevOpsSpec,
  ignored_globs: z.array(z.string()).default([]),
});

const ProfilesFile = z.object({
  profiles: z.record(ProfileSchema).default({}),
});

/** Profile as loaded from the YAML template file. */
export type Profile = z.infer<typeof ProfileSchema> & { name: string };

export const DEFAULT_PROFILES_PATH = path.resolve(__dirname, "../config/profiles.yaml");

function resolvePath(override?: string): string {
  if (override && fs.existsSync(override)) return override;
  const envOverride = process.env.CICD_AGENT_PROFILES_PATH;
  if (envOverride && fs.existsSync(envOverride)) return envOverride;
  return DEFAULT_PROFILES_PATH;
}

const emptyProfile = (name: string): Profile => ({
  name,
  description: "",
  languages: [],
  build: { command: "" },
  test: { command: "" },
  azure_devops: {
    organization: "",
    project: "",
    repository: "",
    default_target_branch: "main",
    pipeline_id: null,
  },
  ignored_globs: [],
});

/** Load all profiles from the YAML template file. Unchanged behaviour. */
export function loadProfiles(profilesPath?: string): Record<string, Profile> {
  const target = resolvePath(profilesPath);
  if (!fs.existsSync(target)) {
    return { default: emptyProfile("default") };
  }
  const text = fs.readFileSync(target, "utf8");
  const raw = YAML.parse(text) ?? {};
  const parsed = ProfilesFile.parse(raw);
  const out: Record<string, Profile> = {};
  for (const [name, p] of Object.entries(parsed.profiles)) {
    out[name] = { name, ...p };
  }
  if (!out.default) out.default = emptyProfile("default");
  return out;
}

/** Get a single YAML template profile by name. */
export function getProfile(name: string, profilesPath?: string): Profile {
  const all = loadProfiles(profilesPath);
  return all[name] ?? all.default ?? emptyProfile(name);
}

// ─── Workspace profiles ────────────────────────────────────────────────────────
//
// "Workspace profiles" are user-created, mutable, and stored in a JSON file
// inside dataDir (never committed to source control).  They extend the YAML
// schema with workspace-specific fields:  local repo path, default/target
// branch, and the ADO PAT (kept out of the YAML for security reasons).
//
// The ADO org / project / repository / pipeline_id fields are deliberately
// duplicated here rather than re-using the nested `azure_devops` spec so that
// the UI can work with a flat, simple object.
// ─────────────────────────────────────────────────────────────────────────────

export interface WorkspaceProfile {
  /** Stable UUID hex identifier */
  id: string;
  /** Human-readable display name */
  name: string;
  createdAt: number;
  updatedAt: number;

  // Local repository
  repoPath: string;
  defaultBranch: string;
  targetBranch: string;

  // Azure DevOps connection (stored per-profile; PAT never written to YAML)
  adoOrgUrl: string;     // e.g. https://dev.azure.com/myorg
  adoProject: string;
  adoRepoName: string;
  adoPat: string;

  // Pipeline (optional)
  adoPipelineId: string;
  adoPipelineName: string;

  // Optional reference to a YAML build/test template
  templateProfile: string;   // name of a Profile from profiles.yaml, or ""

  // Build / test commands — can override the template
  buildCommand: string;
  testCommand: string;
}

export type WorkspaceProfileInput = Omit<WorkspaceProfile, "id" | "createdAt" | "updatedAt">;

type WorkspaceStore = Record<string, WorkspaceProfile>;

function workspaceStorePath(dataDir: string): string {
  return path.join(dataDir, "workspace-profiles.json");
}

function loadWorkspaceStore(dataDir: string): WorkspaceStore {
  const p = workspaceStorePath(dataDir);
  if (!fs.existsSync(p)) return {};
  try {
    return JSON.parse(fs.readFileSync(p, "utf8")) as WorkspaceStore;
  } catch {
    return {};
  }
}

function saveWorkspaceStore(dataDir: string, store: WorkspaceStore): void {
  const p = workspaceStorePath(dataDir);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, JSON.stringify(store, null, 2), "utf8");
}

function nowSec(): number {
  return Math.floor(Date.now() / 1000);
}

/** List all workspace profiles, newest-updated first. */
export function listWorkspaceProfiles(dataDir: string): WorkspaceProfile[] {
  return Object.values(loadWorkspaceStore(dataDir)).sort((a, b) => b.updatedAt - a.updatedAt);
}

/** Get a single workspace profile by id, or null if not found. */
export function getWorkspaceProfile(dataDir: string, id: string): WorkspaceProfile | null {
  return loadWorkspaceStore(dataDir)[id] ?? null;
}

/** Create a new workspace profile and persist it. */
export function createWorkspaceProfile(dataDir: string, data: WorkspaceProfileInput): WorkspaceProfile {
  const store = loadWorkspaceStore(dataDir);
  const id = crypto.randomBytes(8).toString("hex");
  const ts = nowSec();
  const profile: WorkspaceProfile = { ...data, id, createdAt: ts, updatedAt: ts };
  store[id] = profile;
  saveWorkspaceStore(dataDir, store);
  return profile;
}

/** Update an existing workspace profile. Returns null if not found. */
export function updateWorkspaceProfile(
  dataDir: string,
  id: string,
  data: Partial<WorkspaceProfileInput>,
): WorkspaceProfile | null {
  const store = loadWorkspaceStore(dataDir);
  const existing = store[id];
  if (!existing) return null;
  const updated: WorkspaceProfile = { ...existing, ...data, id, updatedAt: nowSec() };
  store[id] = updated;
  saveWorkspaceStore(dataDir, store);
  return updated;
}

/** Delete a workspace profile. Returns false if not found. */
export function deleteWorkspaceProfile(dataDir: string, id: string): boolean {
  const store = loadWorkspaceStore(dataDir);
  if (!store[id]) return false;
  delete store[id];
  saveWorkspaceStore(dataDir, store);
  return true;
}

/**
 * Build the ToolContext extra fields from a workspace profile so that ADO
 * tools receive the correct org / project / repo / PAT automatically.
 */
export function profileToToolExtra(profile: WorkspaceProfile): Record<string, unknown> {
  const orgSlug = profile.adoOrgUrl.replace(/\/$/, "").split("/").pop() ?? profile.adoOrgUrl;
  return {
    ado_org: orgSlug,
    ado_project: profile.adoProject,
    ado_repository: profile.adoRepoName,
    ado_target_branch: profile.targetBranch,
    ado_pat: profile.adoPat,
    ...(profile.adoPipelineId ? { ado_pipeline_id: profile.adoPipelineId } : {}),
  };
}
